04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,97 @@
+const Product = require('../models/product')
+const queryValidator = require('../utils/queryValidator')
+
+const DEFAULT_LIMIT = 10;
+const START_PAGE = 1;
+
+const getAllProductsStatic = async (req, res) => {
+  const products = await Product.find(req.query)
+  res.status(200).json({
+    length: products.length,
+    ...req.query,
+    products
+  })
+}
+const getAllProducts = async (req, res) => {
+  const {
+    name,
+    featured,
+    company,
+    sort,
+    fields,
+    numericFilters,
+  } = req.query
+  let {
+    limit = DEFAULT_LIMIT,
+    page  = START_PAGE
+  } = req.query
+  const queryObject = {}
+
+  if (name) {
+    queryObject.name = { $regex: name, $options: 'i' }
+  }
+  if (featured) {
+    queryObject.featured = (featured === 'true' ? true : false)
+  }
+  if (company) {
+    queryObject.company = company
+  }
+  if (numericFilters) {
+    const operatorsMap = {
+      '>': '$gt',
+      '<': '$lt',
+      '>=': '$gte',
+      '<=': '$lte',
+      '=': '$eq',
+    }
+    const searchOperator = /([<>=]+)/g
+    let filters = numericFilters.replace(searchOperator, match => `-${operatorsMap[match]}-`)
+    console.log(filters)
+    const options = ['price', 'rating']
+    filters = filters.split(',').forEach(item => {
+      const [field, operator, value] = item.split('-')
+      if (options.includes(field)) {
+        queryObject[field] = {
+          [operator]: Number(value)
+        }
+      }
+    })
+    
+  }
+
+  // Create base query
+  const result = Product.find( queryObject )
+  
+  // Sort
+  if (sort) {
+    result.sort( sort.replace(',', ' ') )
```

Comment:
stylistic comment:
This is easily understood, though the general best practice is to avoid nesting like this.  It makes the code more dense and harder to understand.

Reply:
Also, each "action" has the potential for failure.  Nesting like this also makes it harder to troubleshoot.